objects/src/state.rs

```diff
@@ -0,0 +1,84 @@
+use crate::{AccountId, BlockHeader, ChainMmr, Digest, Vec};
+use crypto::merkle::{DefaultMerkleStore, TieredSmt};
+
+/// Structure chain data, used to build necessary openings and to construct [BlockHeader]s.
+#[derive(Default, Debug, Clone)]
+pub struct State {
+    /// An append-only structure used to represent the history of blocks produced for this chain.
+    chain: ChainMmr,
+
+    /// Storage for the historical state of the chain.
+    ///
+    /// This store contains historical data for accounts, nullifiers, notes, and batches.
+    data: DefaultMerkleStore,
+
+    /// History of produced blocks.
+    blocks: Vec<BlockHeader>,
+
+    /// Tree containing the latest [Nullifier] tree.
+    nullifier_db: TieredSmt,
+
+    /// The latest block being produced
+    pending: PendingBlock,
+}
+
+#[derive(Debug, Clone)]
+struct AccountUpdate {
+    account_id: AccountId,
+    new_state: Digest,
+}
+
+#[derive(Default, Debug, Clone)]
+pub struct PendingBlock {
+    /// Account updates in this block
+    accounts: Vec<AccountUpdate>,
+
+    /// Nullifiers used by the consumed notes
+    nullifiers: Vec<Nullifier>,
+}
+
+impl PendingBlock {
+    pub fn account_update(&mut self, account_id: AccountId, new_state: Digest) {
+        self.accounts.push(AccountUpdate {
+            account_id,
+            new_state,
+        });
+    }
+
+    pub fn nullifier(&mut self, nullifier: Nullifier) {
+        self.nullifiers.push(nullifier);
+    }
+}
+
+impl State {
+    /// Get a mutable reference to the pending block
+    pub fn mut_pending(&mut self) -> &mut PendingBlock {
+        &mut self.pending
+    }
+
+    pub fn seal_block(&mut self) -> BlockHeader {
+        let previous = self.blocks.last();
+        let prev_hash = previous.hash().unwrap_or(Digest::default());
+        let block_num = self.blocks.len();
+        let chain_root = self.chain.mmr().accumulator();
+
+        let new = BlockHeader::new(
+            prev_hash,
+            block_num,
+            chain_root,
+            account_root: Digest,
+            nullifier_root: Digest,
+            note_root: Digest,
+            batch_root: Digest,
+            proof_hash: Digest,
+            version: Felt,
+            timestamp: Felt,
+        );
```

Comment:
The question is:

- what is the preferred way of maintaining a account / nullifier / note trees to update the merkle tree and set these values?